source/MRViewer/MRSpaceMouseHandler3dxMacDriver.cpp

```diff
@@ -0,0 +1,249 @@
+                viewer.spaceMouseMove( translate, rotate );
+                glfwPostEmptyEvent();
+            }
+                break;
+        }
+    }
+}
+
+} // namespace
+
+namespace MR
+{
+
+struct SpaceMouseHandler3dxMacDriver::LibHandle
+{
+    void* handle;
+
+#define DEFINE_SYM( name ) name##Func name;
+    DEFINE_SYM( SetConnexionHandlers )
+    DEFINE_SYM( CleanupConnexionHandlers )
+    DEFINE_SYM( RegisterConnexionClient )
+    DEFINE_SYM( SetConnexionClientMask )
+    DEFINE_SYM( SetConnexionClientButtonMask )
+    DEFINE_SYM( UnregisterConnexionClient )
+    DEFINE_SYM( ConnexionClientControl )
+#undef DEFINE_SYM
+
+    bool loadSymbols()
+    {
+#define LOAD_SYM( name ) \
+        {                \
+            ( name ) = (name##Func)dlsym( handle, #name ); \
+            if ( ( name ) == nullptr )                     \
+            {            \
+                spdlog::warn( "Failed to load symbol \"{}\": {}", #name, dlerror() ); \
+                return false;                              \
+            }            \
+        }
+        LOAD_SYM( SetConnexionHandlers )
+        LOAD_SYM( CleanupConnexionHandlers )
+        LOAD_SYM( RegisterConnexionClient )
+        LOAD_SYM( SetConnexionClientMask )
+        LOAD_SYM( SetConnexionClientButtonMask )
+        LOAD_SYM( UnregisterConnexionClient )
+        LOAD_SYM( ConnexionClientControl )
+#undef LOAD_SYM
+        return true;
+    }
+};
+
+SpaceMouseHandler3dxMacDriver::SpaceMouseHandler3dxMacDriver()
+    : lib_( std::make_unique<LibHandle>() )
+{
+    setClientName( "MeshLib" );
+}
+
+SpaceMouseHandler3dxMacDriver::~SpaceMouseHandler3dxMacDriver()
+{
+    if ( lib_->handle != nullptr )
+    {
+        if ( clientId_ )
+        {
+            lib_->UnregisterConnexionClient( clientId_ );
+            // TODO: thread safety?
+            gKnownClientIds.erase( clientId_ );
+        }
+        lib_->CleanupConnexionHandlers();
+        dlclose( lib_->handle );
+    }
+}
+
+void SpaceMouseHandler3dxMacDriver::setClientName( const char* name, size_t len )
+{
+    if ( len == 0 )
+        len = std::strlen( name );
+    assert( len <= 254 );
+    // the name must be converted to the Pascal string format
+    clientName_ = std::make_unique<uint8_t[]>( len + 1 );
+    clientName_[0] = (uint8_t)len;
+    std::memcpy( clientName_.get() + 1, (const uint8_t *)name, len );
+}
+
+bool SpaceMouseHandler3dxMacDriver::initialize()
+{
+    // TODO: better design (e.g. `auto lib = Handle::tryLoad()`)
+
+    static constexpr const auto* c3DconnexionClientPath = "/Library/Frameworks/3DconnexionClient.framework/3DconnexionClient";
+    std::error_code ec;
+    if ( !std::filesystem::exists( c3DconnexionClientPath, ec ) )
+    {
+        spdlog::warn( "3DxWare driver is not installed" );
+        return false;
+    }
+
+    lib_->handle = dlopen( c3DconnexionClientPath, RTLD_LAZY );
+    if ( lib_->handle == nullptr )
+    {
+        spdlog::warn( "Failed to load client library" );
```

Comment:
better write in log something more specific than just "client library", e.g. with "3DxWare"